tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,570 @@
+					Shape: s,
+				},
+			}
+		}
+	case "list(type)", "list(int)", "list(shape)", "list(float)":
+		if lv, ok := v.(*pb.AttrValue_ListValue); ok {
+			return &pb.AttrValue{
+				Value: &pb.AttrValue_List{
+					List: lv,
+				},
+			}
+		}
+	}
+
+	return nil
+}
+
+// Constant Creates a tensor that is added as a constant to the Graph with the
+// specified name.
+func (gr *Graph) Constant(name string, data interface{}) (op *GraphNode, err error) {
+	ts, err := NewTensor(data)
+	if err != nil {
+		return
+	}
+	gr.constants[name] = ts
+
+	return gr.Op("Const", name, nil, "", map[string]interface{}{
+		"dtype": ts.DataType(),
+		"value": ts,
+	})
+}
+
+// matchTypes Matches all the input/output parameters with their corresponding
+// data types specified on the attribues or deducting the data type from other
+// parameters, this method can return an error in case of the matching is not
+// possible, for instance if two input paramters mas have the same data type
+// but one is int and the other float.
+func (gr *Graph) matchTypes(input []*GraphNode, outNode *GraphNode, attrs map[string]interface{}, op *pb.OpDef) (err error) {
+	// Associate the data type tags with the input data types
+	for i, arg := range op.InputArg {
+		inType, inTypeDefined := input[i].outDataTypes[input[i].def.Name]
+		if inTypeDefined && inType != DtInvalid && arg.TypeAttr != "" {
+			attrs[arg.TypeAttr] = inType
+		}
+	}
+	for _, arg := range op.OutputArg {
+		argType := DataType(arg.Type)
+		if arg.TypeAttr != "" && argType != DtInvalid {
+			if inType, defined := attrs[arg.TypeAttr]; defined && inType.(DataType) != argType {
+				return &ErrInputOutputDataTypeMismatch{
+					outDt: argType,
+					inDt:  inType.(DataType),
+				}
+			}
+			attrs[arg.TypeAttr] = arg.Type
+		}
+	}
+
+	// Now assign all the types we got from the inputs/ouputs to their
+	// binded attributes
+	for _, attr := range op.Attr {
+		if attr.Type == "type" {
+			if _, isTypeProvided := attrs[attr.Name]; !isTypeProvided {
+				if inOutDt, inOutDef := attrs[attr.Name]; inOutDef {
+					attrs[attr.Name] = inOutDt
+				} else {
+					if attr.DefaultValue != nil {
+						attrs[attr.Name] = attr.DefaultValue.GetType()
+					}
+				}
+			}
+		}
+	}
+
+	// Assign the corresonding data types from the attributes to the output
+	// params
+	for i, arg := range op.OutputArg {
+		var outDt DataType
+
+		argType := DataType(arg.Type)
+		if argType != DtInvalid {
+			outDt = argType
+		} else {
+			if arg.TypeAttr != "" {
+				if dT, definedDt := attrs[arg.TypeAttr]; definedDt {
+					outDt = dT.(DataType)
+				}
+			}
+		}
+		if len(op.OutputArg) == 1 {
+			outNode.outDataTypes[outNode.def.Name] = outDt
+		} else {
+			// This is a node with more than one output, in this
+			// case the name format is: <name:incremental_id>
+			outNode.outDataTypes[fmt.Sprintf("%s:%d", outNode.def.Name, i)] = outDt
+		}
+	}
+
+	return
+}
+
+// loadAvailableOps Loads all the available operation definitions from local
```

Comment:
I forgot to update this comment, now the definitions are loaded from a .go file that contains the definitions marshalled as string, this file is generated by go generate. The new comment is:
```go
// loadAvailableOps loads all the available operation definitions from a
// constant stored in: proto/tf_ops_def.go that contains a string with all the
// operation definitions
```